src/storage-blob-preview/azext_storage_blob_preview/_validators.py

```diff
@@ -1059,3 +1075,199 @@ def validate_upload_blob(namespace):
         raise InvalidArgumentValueError("usage error: please only specify one of --file and --data to upload.")
     if not namespace.file_path and not namespace.data:
         raise InvalidArgumentValueError("usage error: please specify one of --file and --data to upload.")
+
+
+def _match_path(path, pattern):
+    from fnmatch import fnmatch
+    return fnmatch(path, pattern)
+
+
+def glob_files_locally(folder_path, pattern):
+    """glob files in local folder based on the given pattern"""
+
+    pattern = os.path.join(folder_path, pattern.lstrip('/')) if pattern else None
+
+    len_folder_path = len(folder_path) + 1
+    for root, _, files in os.walk(folder_path):
+        for f in files:
+            full_path = os.path.join(root, f)
+            if not pattern or _match_path(full_path, pattern):
+                yield (full_path, full_path[len_folder_path:])
+
+
+def process_blob_download_batch_parameters(cmd, namespace):
+    """Process the parameters for storage blob download command"""
+    # 1. quick check
+    if not os.path.exists(namespace.destination) or not os.path.isdir(namespace.destination):
+        raise ValueError('incorrect usage: destination must be an existing directory')
+
+    # 2. try to extract account name and container name from source string
+    _process_blob_batch_container_parameters(cmd, namespace)
+
+    # 3. Call validators
+    add_download_progress_callback(cmd, namespace)
+
+
+def process_blob_upload_batch_parameters(cmd, namespace):
+    """Process the source and destination of storage blob upload command"""
+    # 1. quick check
+    if not os.path.exists(namespace.source) or not os.path.isdir(namespace.source):
+        raise ValueError('incorrect usage: source must be an existing directory')
+
+    # 2. try to extract account name and container name from destination string
+    _process_blob_batch_container_parameters(cmd, namespace, source=False)
+
+    # 3. collect the files to be uploaded
+    namespace.source = os.path.realpath(namespace.source)
+    namespace.source_files = [c for c in glob_files_locally(namespace.source, namespace.pattern)]
+
+    # 4. determine blob type
+    if namespace.blob_type is None:
+        vhd_files = [f for f in namespace.source_files if f[0].endswith('.vhd')]
+        if any(vhd_files) and len(vhd_files) == len(namespace.source_files):
+            # when all the listed files are vhd files use page
+            namespace.blob_type = 'page'
+        elif any(vhd_files):
+            # source files contain vhd files but not all of them
+            raise CLIError("""Fail to guess the required blob type. Type of the files to be
```

Comment:
Could we use the more specific error type `ArgumentUsageError` instead of `CLIError`?

Reply:
updated